implement adaptive-p sampler

ddh0 opened the pull request:
This PR implements a new sampler called *adaptive-p* that selects tokens near a configurable target probability over time.

### How it works

The adaptive-p sampler transforms the token probability distribution to favor tokens that fall near a user-configurable probability target. Internally, the sampler maintains an exponential moving average of the *original* probabilities of selected tokens. It uses this, along with the user's set target, to compute an *adapted* target at each sampling step, steering the running average toward the configured target over time. If recent selections have been higher-probability than target, the sampler compensates by temporarily favoring lower-probability tokens, and vice versa.

### Parameters

This sampler exposes two parameters:


| Parameter name | Description                                                                                                | CLI argument          | Valid range | Default value | Notes                                                                                                                                                                                                                        |
| -------------- | ---------------------------------------------------------------------------------------------------------- | --------------------- | ----------- | ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `target`       | Select tokens near this probability                                                                        | `--adaptive-target N` | 0.0 - 1.0   | -1.0          | When set to a negative number, the adaptive probability transform is **disabled**, and instead it just samples normally. Note that since the default value is -1.0, the sampler is disabled by default. This is intentional. |
| `decay`        | Decay value for exponential moving average - lower values are more reactive, higher values are more stable | `--adaptive-decay N`  | 0.0 - 0.99  | 0.90          | Clamped to <=0.99 at init to avoid unbounded accumulation                                                                                                                                                                    |



In most cases, you can just play with `--adaptive-target`. The default decay of 0.9 (for a ~10 token history) works well. A good starting value is `--adaptive-target 0.55`. It is suggested to raise or lower the target in increments of 0.05 as needed.

### Usage notes

adaptive-p selects a token ID rather than just mutating candidates, so it must be last in the sampler chain. It shares this behaviour with some existing samplers like mirostat, dist, and greedy (mirostat being the closest relative).

Only mild truncation before this sampler is recommended. We suggest applying min-p before adaptive-p as the only other active sampler in the chain (optionally with top-k as well).

#### Example usage:

```sh
./build/bin/llama-server -m ~/gguf/my-model.gguf --samplers "top-k;min-p;adaptive-p" --top-k 128 --min-p 0.05 --adaptive-target 0.55
```

### Other notes

This sampler was previously called "power law" in earlier versions of this PR, named for the power law transform we were applying to logits. We are no longer applying the power law transform. We also experimented with gaussian, but ultimately settled on the current formula.

### Acknowledgements

- @MrJackSpade - original "power law" sampler idea and implementation as can be seen [here](https://github.com/MrJackSpade/llama.cpp/) - they continued to help guide me as I worked through the implementation of this sampler. They deserve an **enormous!** amount of credit for the work they put in on this.
- @Geechan for advice on sampler parameters and collaboration during development
- @AesSedai for testing, collaboration, and hosting, as well as coming up with the final name "adaptive-p"

ddh0 committed:
initial commit for branch

ddh0 committed:
simplify constants

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
add params to `struct common_params_sampling`, add reference to PR

ddh0 committed:
explicitly clamp `min_target` and `max_target` to `[0.0, 1.0]`

ddh0 committed:
add args, rename `queue_size` -> `window_size`

ddh0 committed:
improved comments

ddh0 committed:
minor

ddh0 committed:
remove old unused code from algorithm

ddh0 committed:
minor

ddh0 committed:
add power law case to `common_sampler_init`, add sampler name mappings

ddh0 committed:
clarify behaviour when `window_size = 0`

ddh0 committed:
add missing enums

ddh0 commented:
I think this is more or less ready for review now. Also pinging @MrJackSpade in case he'd like to chime in.

ddh0 committed:
remove `target_range` param, make `target == 1` no-op, cleanup code

ddh0 committed:
oops, straggler

pnb reviewed (commented):
This looks very interesting! I wish the original compared to XTC, since the goals seem highly similar.

As an aside, I am curious if there is some way to make it work without selecting a token (i.e., only steps 1-3). I see why token selection is necessary, given the need to save the original probability to the history for the adaptive adjustment part. But, for example, maybe it would suffice instead to save the original probability of the highest-probability token after transforming, regardless of which one is eventually selected by a downstream sampler.

ddh0 commented on src/llama-sampling.cpp:
```diff
@@ -2313,6 +2313,144 @@ struct llama_sampler * llama_sampler_init_dry_testing(int32_t context_size, floa
+
+    // fixed power law transform parameters (from original implementation)
+    const float distribution_width = 0.2f;
+    const float peak_logit_value   = 3.0f;
```
Myself and the original author are discussing the parameters over the next few days, I agree that the current implementation is probably not ideal, which is why I marked it back as draft.

I will post a comment in the main thread with an update once we've got it more figured out. Thank you!

ddh0 committed:
add missing parameters in `server-task.cpp`

ddh0 committed:
copy from author

ref:
https://gist.github.com/MrJackSpade/9be99c7efbba7b95a41377e123b7b069

ddh0 committed:
remove old debug log, style nit

ddh0 committed:
fix compiler warning, add commented-out logging per token

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
re-write + change parameters + simplify

ddh0 committed:
oops forgot args.cpp

ddh0 committed:
fix leftover `window_size`

ddh0 committed:
add missing values to `common_params_sampling::print()`

ddh0 committed:
with logging

ddh0 committed:
does this fix it?

ddh0 committed:
no, but does this?

ddh0 committed:
update default decay

ddh0 committed:
Merge branch 'master' into power-law-sampler

ddh0 committed:
optimize

ddh0 committed:
fix bad merge

my git skills are lacking

ddh0 committed:
silence `missing initializer for member`

ddh0 committed:
update default decay to 0.9

ddh0 committed:
fix logging

ddh0 committed:
format (double)

ddh0 committed:
add power law to the new `samplers` vector

ddh0 committed:
log sampler init values

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
improve logging messages in llama_sampler_power_law

ddh0 committed:
remove extraneous logging

ddh0 committed:
simplify target computation

last commit with debug logging!

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
remove debug logging, explicitly clamp params at init

ddh0 commented:
@pnb I've basically re-done the entire PR since your last comment, as well as updated the top comment with a much clearer explanation. Let me know if I can clear anything up.

Geechan commented:
This is a fantastic sampler for creative tasks, and is truly a game changer in this regard. It's difficult to understand how effective it is until you try it for yourself.

I've found a target value of 0.4-0.7 to be excellent for creative tasks, with higher values for more deterministic tasks. It manages to do what XTC does without many of the pitfalls behind that sampler - a self correcting, dynamic algorithm basically keeps the sampler in check much better than setting a random chance to apply top truncation to. A lot of so called 'AI slop' is heavily over-represented in the top tokens, and so this sampler really helps a model shine in the still strong and coherent mid range while not drifting too far from established probabilities and the natural distribution charts of models (unlike adjusting temperature).

I hope to see this merged!

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
add `use_power_law` flag + logic, minor cleanup

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
update `power-law` -> `adaptive-p`

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
fix cold start EMA

- `ctx->weighted_sum` is now initialized and reset to `target / (1.0f -
clamped_decay)`
- `ctx->total_weight` is now initialized and reset to `1.0f / (1.0f -
clamped_decay)`

this fixes a "cold start" problem with the moving average

ddh0 commented:
Barring any changes requested by reviewers/maintainers, I believe this implementation to be correct and finalized at this point. Just waiting on #17004.

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
update `SHARPNESS` constant to `10.0f`

ddh0 committed:
minor style fixes

no functional changes

ddh0 committed:
minor style fixes cont.

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
update `llama_sampler_adaptive_p_i` for backend sampling (ref: #17004)

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 commented on src/llama-sampling.cpp:
```diff
@@ -3293,6 +3293,142 @@ struct llama_sampler * llama_sampler_init_dry_testing(int32_t context_size, floa
+
+    // update EMA with the original probability of the selected token
+    ctx->weighted_sum = ctx->original_probs[idx] + ctx->decay * ctx->weighted_sum;
+    ctx->total_weight = 1.0f + ctx->decay * ctx->total_weight;
```
> Some sampled tokens could get discarded

Could you please elaborate on this? I'm not sure I understand

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
separate into `apply` + `accept` functions

Beinsezii commented:
> I suspect our softmax + transform + softmax is effectively "reviving" some tokens that should have been completely "killed" by min-p in this case.

mathematically softmax should handle a proper value like -inf
```
>>> softmax([1, 2])
(0.2689414213699951, 0.7310585786300048)
>>> softmax([-float('inf'), -float('inf'), 1, 2])
(0.0, 0.0, 0.2689414213699951, 0.7310585786300048)
```
so the -inf masked tokens will eat the offset while retaining exact zero after sm

MrJackSpade commented:
Pardon my Claude, but:

The issue occurs because the adaptive-p transformation loop processes all candidates in `cur_p->size` before the second softmax:

```cpp
// Lines 3356-3359: Transform ALL candidates in the array
for (size_t i = 0; i < cur_p->size; ++i) {
    float dist = std::abs((cur_p->data[i].p - adapted_target) * INV_WIDTH);
    cur_p->data[i].logit = PEAK_LOGIT_VALUE - SHARPNESS * dist * dist / (1.0f + dist);
}
```

With **CPU sampling**, samplers like top-k or top-p truncate the candidate array by setting `cur_p->size` to the number of valid tokens. The transformation loop only sees these valid tokens. There are no `-inf` logits to worry about because they were physically excluded from the array.

With **backend (GPU) sampling**, the array is not truncated. Instead, `cur_p->size` remains at `n_vocab`. Tokens that should be suppressed are signaled as invalid by having `-inf` logits. The problem is:

1. The **first softmax** (line 3337) converts `-inf` logits to `p ≈ 0.0`
2. The **transform loop** (lines 3356-3359) then calculates new logits for these tokens:
   ```cpp
   // For a token with p ≈ 0.0 and adapted_target = 0.6:
   dist = |0.0 - 0.6| / 0.3 = 2.0
   logit = 5.0 - 10.0 * 4.0 / 3.0 ≈ -8.33  // NOT -inf!
   ```
3. The **second softmax** (line 3362) now sees these as valid candidates with `logit ≈ -8.33`, giving them non-zero probability and allowing them to be sampled

In short, the adaptive probability transform scores tokens based on how close their original probability is to a target. Tokens with `p ≈ 0` get a low score, but that score is finite (`≈ -8.33`), not `-inf`. When all tokens in the vocabulary pass through this transform (as happens with backend sampling) previously suppressed tokens "resurrect" with small but non-zero probabilities.

MrJackSpade commented:
I would like to see whether or not the backend min-p implementation is an error or not. 

Both the frontend min-p and the backend top-k properly set the "size" of the resulting collection, which means the current backend min-p doesn't behave as expected when compared to not only its own alternate implementation, but also other backend samplers. It's not consistent in its current behavior.

It would be preferable if nothing needed to change to account for it, even if only to avoid all of the testing that needs to be redone as part of it.

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
`pending_token_idx`: switch from `llama_token` to `int32`

functionally identical (`llama.h` has `typedef int32_t llama_token;`),
but its more correct now

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 commented:
Hello @danbev and @ggerganov - what should be the next steps here? Thank you.

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
don't transform logits <= -1e9f

ddh0 commented:
> thing I worry is someone else will pick a different magic number than 1e9 in the future and then adaptive-p breaks again. that's why I think adaptive-p should calculate its own zero point against -inf or -1e9 or whatever.

Yes, I'm not sure. `-1e9f` is fine for now as far as I'm aware but I'll wait for @ggerganov to give his input on this. Maybe we can have a file-level constant for this threshold.

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
fix masking in backend top-p, min-p

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ggerganov commented on src/llama-sampling.cpp:
```diff
@@ -3293,6 +3286,166 @@ struct llama_sampler * llama_sampler_init_dry_testing(int32_t context_size, floa
+
+    // softmax and store the original probabilities
+    llama_sampler_softmax_impl(cur_p, false);
+    ctx->original_probs.resize(cur_p->size);
```
The `llama_sampler_softmax_impl(cur_p, false);` call can be deduplicated here - call once before the `if`

ggerganov commented on src/llama-sampling.cpp:
```diff
@@ -3293,6 +3286,166 @@ struct llama_sampler * llama_sampler_init_dry_testing(int32_t context_size, floa
+    float    decay,
+    uint32_t seed
+) {
+    auto seed_cur = get_rng_seed(seed);
```
This `seed_cur` logic breaks when the `seed` is `LLAMA_DEFAULT_SEED`. In that case we generate a random seed based on the time. But with this implementation, if you clone the sampler, it will inherit the "seed_cur" of the source sampler, instead of generating a new seed.

See how the `dist` sampler maintains a `seed_cur` in its context and replicate the logic here.

ggerganov reviewed (approved):
After addressing the minor comments we can merge

ddh0 committed:
address review comments

ddh0 committed:
typo in comments `RND` -> `RNG`

z80maniac commented:
I think this sampler needs to be described in the server's README. Otherwise the users will have no idea how to use it or that it even exists in the first place.

ddh0 committed:
add docs

ddh0 committed:
add recommended values in completion docs

ddh0 committed:
address PR feedback

ddh0 commented on src/llama-sampling.cpp:
```diff
@@ -3293,6 +3286,168 @@ struct llama_sampler * llama_sampler_init_dry_testing(int32_t context_size, floa
+    ctx->pending_token_id  = LLAMA_TOKEN_NULL;
+    ctx->pending_token_idx = -1;
+}
+
```
Oops. I addressed this in `40fd48f` just now

ddh0 committed:
remove trailing whitespace (for CI `editorconfig`)

timwu commented on common/sampling.cpp:
Should adaptive_p be present in this map as well to allow conversion from char to sampler?

ddh0 committed:
Merge branch 'ggml-org:master' into power-law-sampler

ddh0 committed:
add to adaptive-p to `common_sampler_types_from_chars`
